### DISCUSSION TITLE
How to define type which could differ due to ImportError

### DISCUSSION THREAD

**Discussion by BrandonNav** [microsoft/pyright#4160] [Q&A] [answered]
I have code which imports an Enum but that import might not be available, so a try/except ImportError is setup.
In the except block, I define a fallback enum with the same name so the code still functions. 

Pyright gives an error that the types are not the same; i believe since the type is assumed to be from the first declaration.

**Question**
I am wondering how to define the union type so my functions can take either enum.

**Context**
pyright: 1.1.278
Run on cli and in vscode

**Example Code**

```py
from enum import IntEnum

try:
    from http import HTTPStatus        
except ImportError:
    class HTTPStatus(IntEnum):
        NOT_FOUND = 404
        
def get_status_code_as_str(status: HTTPStatus) -> str:
    return str(status.value)
```

**Error**
```py
/private/tmp/test/ht.py
  /private/tmp/test/ht.py:4:22 - error: Expression of type "Type[HTTPStatus]" cannot be assigned to declared type "Type[HTTPStatus]"
    "http.HTTPStatus" is incompatible with "ht.HTTPStatus"
    Type "Type[HTTPStatus]" cannot be assigned to type "Type[HTTPStatus]" (reportGeneralTypeIssues)
```

**Some things I tried**
I can change the redefined `HTTPStatus(Enum)` to 
```py
HTTPStatus = type('HTTPStatus', (IntEnum,), {"NOT_FOUND": 404})
```

But I believe that is just working around the typechecker.

I also thought I could create a typealias by renaming the enums, i.e.
```py
if TYPE_CHECKING:     
    HTTPStatus: TypeAlias = HTTPStatusImport | HTTPStatusRedefined
```

But the typechecker correctly tells me that HTTPStatusImport is unbound.

**Comment by erictraut** [comment, marked as answer]
I recommend against attempting to create a union. Stick to one type or the other and use that for static type checking.

```python
from enum import IntEnum
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    class HTTPStatus(IntEnum):
        NOT_FOUND = 404
else:
    try:
        from http import HTTPStatus
    except ImportError:
        class HTTPStatus(IntEnum):
            NOT_FOUND = 404
```

Or you could use a conditional check based on `sys.version`:

```python
import sys

if sys.version_info >= (3, 11):
    from http import HTTPStatus        
else:
    from enum import IntEnum
    class HTTPStatus(IntEnum):
        NOT_FOUND = 404
```

**Reply by BrandonNav** [reply]
Ah, yes that makes sense. Thank you
